ocean_lib/common/ddo/credentials.py

```diff
@@ -0,0 +1,112 @@
+        entries = self.asset._credentials.get(access_class, [])
+
+        for entry in entries:
+            if entry["type"] == "address":
+                return [val.lower() for val in entry["values"]]
+
+        return []
+
+    def requires_credential(self):
+        """Checks whether the asset requires an address credential."""
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        return allowed_addresses or denied_addresses
+
+    def validate_access(self, credential=None):
+        """Checks a credential dictionary against the address allow/deny lists."""
+        address = simplify_credential_to_address(credential)
+
+        allowed_addresses = self.get_addresses_of_class("allow")
+        denied_addresses = self.get_addresses_of_class("deny")
+
+        if not address and not self.requires_credential():
+            return ConsumableCodes.OK
+
+        if allowed_addresses and address.lower() not in allowed_addresses:
+            return ConsumableCodes.CREDENTIAL_NOT_IN_ALLOW_LIST
+
+        if not allowed_addresses and address.lower() in denied_addresses:
+            return ConsumableCodes.CREDENTIAL_IN_DENY_LIST
+
+        return ConsumableCodes.OK
+
+    def add_address_to_access_class(self, address, access_class="allow"):
+        """Adds an address to an address list (either allow or deny)."""
+        address = address.lower()
+
+        if not self.asset._credentials or access_class not in self.asset._credentials:
+            self.asset._credentials[access_class] = [
+                {"type": "address", "values": [address]}
+            ]
+            return
+
+        address_entries = self.get_address_entry_of_class(access_class)
+
+        if not address_entries:
+            self.asset._credentials[access_class].append(
+                {"type": "address", "values": [address]}
+            )
+            return
+
+        address_entry = address_entries[0]
+        lc_addresses = self.get_lc_addresses_from_entry(address_entry)
+
+        if address not in lc_addresses:
+            lc_addresses.append(address)
+
+        address_entry["values"] = lc_addresses
+
+    def remove_address_from_access_class(self, address, access_class="allow"):
+        """Removes an address from an address list (either allow or deny)i."""
+        address = address.lower()
+
+        if not self.asset._credentials or access_class not in self.asset._credentials:
+            return
+
+        address_entries = self.get_address_entry_of_class(access_class)
+
+        if not address_entries:
+            return
+
+        address_entry = address_entries[0]
+        lc_addresses = self.get_lc_addresses_from_entry(address_entry)
+
+        if address not in lc_addresses:
+            return
+
+        lc_addresses.remove(address)
+        address_entry["values"] = lc_addresses
+
+    def get_address_entry_of_class(self, access_class="allow"):
```

Comment:
I think `get_address_entry_of_class` should extract the 0th entry before returning because there should only ever be 1 address entry in a given access class list.

Also I just realized that I mixed up the docstrings between `get_address_entry_of_class` and `get_lc_addresses_from_entry`. So I decided to just re-write it to be a little more clear.

```python
    def get_address_entry_of_class(self, access_class="allow"):
        """Get address credentials entry of the specified access class. access_class = "allow" or "deny"."""
        entries = self.asset._credentials.get(access_class, [])
        address_entries = [entry for entry in entries if entry["type"] == "address"]
        return address_entries[0] if address_entries else None
```